src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -232,18 +232,20 @@ private sealed class FormatterVisitor : TypeShapeVisitor
     {
         private readonly TypeDictionary formatters = new()
         {
-            { typeof(string), new Formatter<string>((ref MessagePackReader reader, MessagePackSerializerOptions options) => reader.ReadString(), (ref MessagePackWriter writer, string? value, MessagePackSerializerOptions options) => writer.Write(value)) },
-            { typeof(int), new Formatter<int>((ref MessagePackReader reader, MessagePackSerializerOptions options) => reader.ReadInt32(), (ref MessagePackWriter writer, int value, MessagePackSerializerOptions options) => writer.Write(value)) },
+            { typeof(string), new PrimitiveFormatter<string>((ref MessagePackReader reader) => reader.ReadString(), (ref MessagePackWriter writer, string? value) => writer.Write(value)) },
+            { typeof(int), new PrimitiveFormatter<int>((ref MessagePackReader reader) => reader.ReadInt32(), (ref MessagePackWriter writer, int value) => writer.Write(value)) },
         };
 
         internal IMessagePackFormatter<T?> GetFormatter<T>(ITypeShape<T> typeShape)
         {
-            return formatters.GetOrAdd<IMessagePackFormatter<T?>>(typeShape, this, box => new Formatter<T?>(box.Result.Deserialize, box.Result.Serialize));
+            return formatters.GetOrAdd<IMessagePackFormatter<T?>>(typeShape, this, box => box.Result switch
```

Comment:
The delayed value factory is a late binding mechanism used when a recursive type is encountered. The idea is that it's creating a wrapper for `box.Result` which will _eventually_ contain a reference to the final computed value once building has completed. 

For this to work, evaluating `box.Result` needs to be delayed until an actual serialization operation is run. Here's how it's done for CBOR:

https://github.com/eiriktsarpalis/typeshape-csharp/blob/c8c2c397beb472e716b447bb460dea626811ba18/src/TypeShape.Examples/CborSerializer/Converters/DelayedCborConverter.cs#L6-L13

And here's how the object cloner does it:

https://github.com/eiriktsarpalis/typeshape-csharp/blob/main/src/TypeShape.Examples/Cloner/Cloner.cs#L62